File: testsuite/cppkeepbodies/more11features.cpp

#def static_assert(x, y)

struct Double
{
    static_assert(std::numeric_limits<double>::is_iec559
               && std::numeric_limits<double>::digits == 53
               && std::numeric_limits<double>::max_exponent == 1024,
        "IEEE-754 double-precision implementation required");

    using value_type = double;
    using bits_type = uint64_t;

    static constexpr int32_t   SignificandSize = std::numeric_limits<value_type>::digits; // = p   (includes the hidden bit)
    static constexpr int32_t   ExponentBias    = std::numeric_limits<value_type>::max_exponent - 1 + (SignificandSize - 1);
    static constexpr bits_type MaxIeeeExponent = bits_type{2 * std::numeric_limits<value_type>::max_exponent - 1};
    static constexpr bits_type HiddenBit       = bits_type{1} << (SignificandSize - 1);   // = 2^(p-1)
    static constexpr bits_type SignificandMask = HiddenBit - 1;                           // = 2^(p-1) - 1
    static constexpr bits_type ExponentMask    = MaxIeeeExponent << (SignificandSize - 1);
    static constexpr bits_type SignMask        = ~(~bits_type{0} >> 1);

    bits_type bits;

    explicit Double(bits_type bits_) : bits(bits_) {}
    explicit Double(value_type value) : bits(ReinterpretBits<bits_type>(value)) {}

};

#include <cstdint>
#include <vector>

int main(){
  std::vector<int64_t> foo(10);
  return 0;
}

#define MYIGNORE
#define MYCDECL   __cdecl

int test1() { int x = 1; return (x); }
int MYCDECL test2() { int x = 2; return (x); }

#if defined(MYIGNORE)
  int myVar;

  MYIGNORE int test3() {
    myVar = test1();
	myVar = myVar + test2();
	return(myVar);
  }

#endif

#ifdef MYIGNORE

  MYIGNORE int test4() {
    myVar = test1();
	myVar = myVar + test2();
	return(myVar);
  }

#endif

#ifdef DEBUG
#  define OUT(x) printf("%s\n", x)
#else
#  define OUT(x)
#endif

// bug #190

#def Q_DISABLE_COPY(MyClass) \
    MyClass(const MyClass &) = delete; \
    MyClass &operator=(const MyClass &) = delete;

#define Q_CORE_EXPORT

class Q_CORE_EXPORT QObjectData
{
    Q_DISABLE_COPY(QObjectData)
public:
    QObjectData() = default;
    virtual ~QObjectData() = 0;
    QObject *q_ptr;
    QObject *parent;
    QObjectList children;

    uint isWidget : 1;
    uint blockSig : 1;
    uint wasDeleted : 1;
    uint isDeletingChildren : 1;
    uint sendChildEvents : 1;
    uint receiveChildEvents : 1;
    uint isWindow : 1; // for QWindow
    uint deleteLaterCalled : 1;
    uint unused : 24;
    int postedEvents;
    QDynamicMetaObjectData *metaObject;
    QBindingStorage bindingStorage;
    QMetaObject *dynamicMetaObject() const override final;
};

// C++ lambdas

auto ex1 = [] (int x) { std::cout << x << '\n'; };

auto ex2 = [] ()   { code; };

auto ex3 = [](float f, int a) { return a*f; };
auto ex4 = [](MyClass t) -> int { auto a = t.compute(); return a; };
auto ex5 = [](int a, int b) { return a < b; };

auto myLambda = [](int a) -> double { return 2.0 * a; };

auto myLambda = [](int a) mutable { std::cout << a; };

auto baz = [] () {
    int x = 10;
    if ( x < 20)
        return x * 1.1;
    else
        return x * 2.1;
};

int x = 1, y = 1;
[&]() { ++x; ++y; }(); // <-- call ()


int main() {
    int x = 10;
    int y = 11;
    // Captures With an Initializer
    auto foo = [z = x+y]() { std::cout << z << '\n'; };
    foo();

    std::unique_ptr<int> p(new int{10});
    auto foo = [x=10] () mutable { ++x; };
    auto bar = [ptr=std::move(p)] {};
    auto baz = [p=std::move(p)] {};
}

// decltype
int i;

decltype(i+3) j;

typedef std::function<void(void)> Foo;

// bug #78
for (int i = 0; i < 44; i++) {
  if (a[i]) continue;
  print(a[i]);
}

// smart def vs define heuristic:
#define for_each(x) for(int i = 0; i < x; ++i)

#define other(x) for(int i = 0; i < x; ++i) printf(i);

template <typename T>
class Foo {
public:
        Foo(){
          for_each(89) printf(i);
          other(13);

        };
};

// bug #59

enum class Color { red, green = 20, blue };

class MyClass {
  Color color;
public:
  void (*warning)(const char*, ...);    // <- this fails!!
  void *warning(const char*, ...);


  T&       value() &;
  T&&      value() &&;
  T const& value() const&;
};

void* MyClass::warning(const char*, ...) {
  int bodyHere;
  switch (this->color) {
    using enum Color;
    case red: ;
    case green:
      bodyHere = 123;
    case blue: ;
  }

  if (auto f = (5+6); f != 0) {
    printf("I love syntactic sugar!\n");
  }
}

File: testsuite/results/more11features.nim
type
  Double* {.bycopy.} = object
    bits*: bits_type

  value_type* = cdouble
  bits_type* = uint64_t

const                       ##  = p   (includes the hidden bit)
  SignificandSize*: int32_t = digits[value_type]

const
  ExponentBias*: int32_t = max_exponent[value_type] - 1 + (SignificandSize - 1)

const
  MaxIeeeExponent*: bits_type = bits_type(2 * max_exponent[value_type] - 1)

const                       ##  = 2^(p-1)
  HiddenBit*: bits_type = bits_type(1) shl (SignificandSize - 1)

const                       ##  = 2^(p-1) - 1
  SignificandMask*: bits_type = HiddenBit - 1

const
  ExponentMask*: bits_type = MaxIeeeExponent shl (SignificandSize - 1)

const
  SignMask*: bits_type = not (not bits_type(0) shr 1)

proc constructDouble*(bits_: bits_type): Double {.constructor.} =
  discard

proc constructDouble*(value: value_type): Double {.constructor.} =
  discard

proc main*(): cint =
  var foo: vector[int64_t] = vector[int64_t](10)
  return 0

proc test1*(): cint =
  var x: cint = 1
  return x

proc test2*(): cint {.cdecl.} =
  var x: cint = 2
  return x

var myVar*: cint

proc test3*(): cint =
  myVar = test1()
  myVar = myVar + test2()
  return myVar

proc test4*(): cint =
  myVar = test1()
  myVar = myVar + test2()
  return myVar

when defined(DEBUG):
  template OUT*(x: untyped): untyped =
    printf("%s\n", x)

else:
  discard
##  bug #190

type
  QObjectData* {.bycopy.} = object
    q_ptr*: ptr QObject
    parent*: ptr QObject
    children*: QObjectList
    isWidget* {.bitsize: 1.}: uint
    blockSig* {.bitsize: 1.}: uint
    wasDeleted* {.bitsize: 1.}: uint
    isDeletingChildren* {.bitsize: 1.}: uint
    sendChildEvents* {.bitsize: 1.}: uint
    receiveChildEvents* {.bitsize: 1.}: uint
    isWindow* {.bitsize: 1.}: uint ##  for QWindow
    deleteLaterCalled* {.bitsize: 1.}: uint
    unused* {.bitsize: 24.}: uint
    postedEvents*: cint
    metaObject*: ptr QDynamicMetaObjectData
    bindingStorage*: QBindingStorage


proc constructQObjectData*(): QObjectData {.constructor.}
proc destroyQObjectData*(this: var QObjectData)
proc dynamicMetaObject*(this: QObjectData): ptr QMetaObject {.noSideEffect.}
##  C++ lambdas

var ex1*: auto = (proc (x: cint): auto =
  cout shl x shl '\n')

var ex2*: auto = (proc (): auto =
  code)

var ex3*: auto = (proc (f: cfloat; a: cint): auto =
  return a * f)

var ex4*: auto = (proc (t: MyClass): cint =
  var a: auto = t.compute()
  return a)

var ex5*: auto = (proc (a: cint; b: cint): auto =
  return a < b)

var myLambda*: auto = (proc (a: cint): cdouble =
  return 2.0 * a)

var myLambda*: auto = (proc (a: cint): auto =
  cout shl a)

var baz*: auto = (proc (): auto =
  var x: cint = 10
  if x < 20:
    return x * 1.1
  else:
    return x * 2.1
  )

var
  x*: cint = 1
  y*: cint = 1

(proc (): auto =
  inc(x)
  inc(y))()
##  <-- call ()

proc main*(): cint =
  var x: cint = 10
  var y: cint = 11
  ##  Captures With an Initializer
  var foo: auto = (proc (): auto =
    cout shl z shl '\n')
  foo()
  var p: unique_ptr[cint] = unique_ptr[cint](new(int(10)))
  var foo: auto = (proc (): auto =
    inc(x))
  var bar: auto = (proc (): auto = discard )
  var baz: auto = (proc (): auto = discard )

##  decltype

var i*: cint

var j*: typeof(i + 3)

type
  Foo* = function[proc ()]

##  bug #78

var i*: cint = 0

while i < 44:
  if a[i]:
    inc(i)
    continue
  print(a[i])
  inc(i)
##  smart def vs define heuristic:

template other*(x: untyped): void =
  var i*: cint = 0
  while i < x:
    printf(i)
    inc(i)

type
  Foo*[T] {.bycopy.} = object


proc constructFoo*[T](): Foo[T] {.constructor.} =
  var i: cint = 0
  while i < 89:
    printf(i)
    inc(i)
  other(13)

##  bug #59

type
  Color* = enum
    red, green = 20, blue


type
  MyClass* {.bycopy.} = object
    color*: Color
    warning*: proc (a1: cstring) {.varargs.} ##  <- this fails!!


proc warning*(this: var MyClass; a2: cstring): pointer {.varargs.}
proc value*(this: var MyClass): var T
proc value*(this: var MyClass): var T
proc value*(this: MyClass): T {.noSideEffect.}
proc warning*(this: var MyClass; a2: cstring): pointer {.varargs.} =
  var bodyHere: cint
  case this.color
  of red:
    ## ignored statement
  of green:
    bodyHere = 123
  of blue:
    ## ignored statement
  if (
    var f: auto = (5 + 6)
    f != 0):
    printf("(I love syntactic sugar!\\n")
    printf("(I love syntactic sugar!\\n")

type
  Coord* {.bycopy.} = object
    x*: cint
    y*: cint
    z*: cint


const
  arr*: array[3, Coord] = [Coord(x: 1, y: 2, z: 3), Coord(x: 4, y: 5, z: 6),
    Coord(x: 7, y: 8, z: 9)]

##  test 'noexcept':

proc stuff*()
proc raisesNothing*(): cint {.raises: [].} =
  discard
